### PR TITLE
[MoreUnit Light] Sourcefile and testfile should be allowed to have different file extensions.

### PR DESCRIPTION
**Pull request by niecore**
- allows the user to use heterogen file extensions for test and source files.
- can be configured in the preferences screen.
- improved preferences group handling

### CODE REVIEW

**Comment by gianasista** [comment]
The IntelliJ plugin works with different file extension already (HelloWorld.java switches to HelloWorldTest.scala and other extensions like HelloWorldTest.test).
Anyway this pull request is an improvement.
@ndemengel: What do you think?

**Comment by ndemengel** [comment]
Hi,

I apologize for not having answered. Of course it would be a welcome improvement! Thank you @niecore!
Unfortunately I really don't have enough time to review the code and/or test the feature, so I hope @gianasista or anyone else will found that time :-)

**Comment by niecore** [comment]
@ndemengel Okay, but anyhow there is no need to hurry.

**Comment by gianasista** [comment]
@niecore Do all the test from org.moreunit.core.test run on your computer?
At the moment 3 test are failing on my machine (but don't have the time now to investigate further).

**Comment by gianasista** [comment]
It's my workspace's fault that the tests do not run, there is something wrong when reading the file association prefs. Something is messed up on my machine. It still insists opening clj-files with the system default although it is configured in the file association prefs.

**Comment by gianasista** [comment]
Can anyone give feedback that all tests run on his machine?
I already changed system default file association but when running JumpActionHandlerTest (from org.moreunit.core.test) files with *.clj extension are opened with Atom (did not check what ExtensionRegistry does with the language extension in the //when-part, but for me, there seems to be a problem only while running plugin tests - on master branch too). (fresh eclipse installation, fresh moreunit clone, fresh junit-testworkspace...)

**Comment by niecore** [comment]
@gianasista all 215 are green in in my workspace on my machine. (sorry for the late reply)

**Comment by apupier** [comment]
@niecore it seems that you have different end of line for some specific files, it is making harder to review the PR. Can you have a look at it please?

**Comment by apupier** [comment]
I have errors for several tests, all of them have this error:
image

{noformat}
Tests in error: 
  should_update_test_package_prefix_when_preferences_change(org.moreunit.settings.PreferencesTest): Could not find widget matching: (of type 'Text' and with label (with mnemonic 'Test package prefix:'))
  should_update_test_superclass_when_preferences_change(org.moreunit.settings.PreferencesTest): Could not find menu bar for shell: Shell with text {Could Not Accept Changes }
  should_update_test_type_when_preferences_change(org.moreunit.settings.PreferencesTest): Could not find menu bar for shell: Shell with text {Could Not Accept Changes }
  should_enable_extended_method_search_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_source_folder_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_package_suffix_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_method_prefix_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_name_template_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_method_content_when_preferences_change(org.moreunit.settings.PreferencesTest): The widget was null.
  should_update_test_package_prefix_when_preferences_change(org.moreunit.settings.PropertiesTest): Could not find menu bar for shell: Shell with text {Could Not Accept Changes }
  should_update_test_superclass_when_preferences_change(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_update_test_method_content_when_property_changes(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_update_test_prefixes_when_preferences_change(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_enable_extended_method_search_when_preferences_change(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_update_test_package_suffix_when_preferences_change(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_DisabledUI_when_no_project_specific_settings_used(org.moreunit.settings.PropertiesTest): Could not find widget matching: (of type 'Button' and with mnemonic 'Use project specific settings' and with style 'SWT.CHECK')
  should_update_test_method_prefix_when_property_changes(org.moreunit.settings.PropertiesTest): Could not find menu bar for shell: Shell with text {Could Not Accept Changes }
  should_update_test_type_when_property_changes(org.moreunit.settings.PropertiesTest): Could not find menu bar for shell: Shell with text {Could Not Accept Changes }
{noformat}

**Comment by niecore** [comment]
I see. I will investigate into this.

**Comment by niecore** [comment]
Updated the branch so that now also the SWT-Bot tests are running.

**Comment by niecore** [comment]
Is there anything I can do to help making reviewing this PR easier for you?

File: org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        {
+            testFileExt.setText("");
+            
+        }
+
+        testFileExt.setEnabled(extEnable);
+        srcFileExt.setEnabled(extEnable);
+        checkBox.setSelection(extEnable);
+        
+        checkBox.setText("Heterogen file extensions (if unchecked the same file extension for source and test files assumed");
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java]: missing closing parenthesis at the end.

File: org.moreunit.core/src/org/moreunit/core/matching/FileMatcher.java [hunk 2]
```diff
@@ -1,82 +1,96 @@
...
+            ext = file.getExtension();
+        }
+        
+        Pattern fileNamePattern2 = createFileNamePattern(ext, filePatterns);
+        searchEngine.searchFiles(searchFolder, fileNamePattern2, matchCollector);
+    }
+
+    private String createPreferredFileName(FileNameEvaluation evaluation)
+    {
+        String prefferedFileName = "";
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/matching/FileMatcher.java]: typo preffered -> preferred

File: org.moreunit.core/src/org/moreunit/core/matching/FileNameEvaluation.java [hunk 3]
```diff
@@ -140,4 +143,14 @@ public String toString()
                              preferredCorrespondingFilePatterns, StringConstants.NEWLINE, //
                              otherCorrespondingFilePatterns, StringConstants.NEWLINE);
     }
+
+    public String getCorrespondingExtenstion()
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/matching/FileNameEvaluation.java]: typo getCorrespondingExtenstion -> getCorrespondingExtension

File: org.moreunit.core/src/org/moreunit/core/preferences/FolderPatternGroup.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        createFolderTemplateFields(folderTplGroup);
+        createFolderTplExplanations(folderTplGroup);
+        
+    }
+
+    @Override
+    public void forceFocus()
+    {
+        // TODO Auto-generated method stub
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/FolderPatternGroup.java]: Can you remove the TODO please?
If nothing, to put inside worth specifying a comment "// no action" or something like that

File: org.moreunit.core/src/org/moreunit/core/preferences/FolderPatternGroup.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    @Override
+    public String getError()
+    {
+        return validateTestFolderTemplate();
+    }
+
+    @Override
+    public String getWarning()
+    {
+        // TODO Auto-generated method stub
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/FolderPatternGroup.java]: Can you remove the TODO please?

File: org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+import org.moreunit.core.ui.Composites;
+import org.moreunit.core.ui.ExpandableCompositeContainer;
+import org.moreunit.core.ui.ExpandableCompositeContainer.ExpandableContent;
+import org.moreunit.core.ui.FileNamePatternDemo;
+import org.moreunit.core.ui.Labels;
+import org.moreunit.core.ui.LayoutData;
+
+public class FileExtGroup implements GenericPreferencesGroup
+{
+    private static final int EXPLANATION_WIDTH_HINT = 350;
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java]: this constant is not used. Please remove it or use it somewhere.

File: org.moreunit.core/src/org/moreunit/core/preferences/TestFileNamePatternGroup.java [hunk 7]
```diff
@@ -42,14 +44,21 @@ private TestFileNamePatternGroup(Composite parent, ExpandableCompositeContainer
...
 
+    @Override
+    public void createContents()
+    {
         testFileTemplateField = createFileTemplateField(fileTplGroup);
         wordSeparatorField = createWordSeparatorField(fileTplGroup, forceCamelCase);
         createFileTplExplanations(fileTplGroup);
-        createOverviewArea(fileTplGroup);
-    }
+        createOverviewArea(fileTplGroup);
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/TestFileNamePatternGroup.java]: optional: if you can clean the added spaces

File: org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
+package org.moreunit.core.preferences;
+
+import static org.moreunit.core.util.Strings.countOccurrences;
+
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.ModifyEvent;
```
Comment (apupier) [inline org.moreunit.core/src/org/moreunit/core/preferences/FileExtGroup.java]: several imports are unused

**Comment by apupier** [comment]
> Is there anything I can do to help making reviewing this PR easier for you?

I'm very sorry for this very late review.

**Comment by apupier** [comment]
@niecore I provided a PR on your branch for several of my comments https://github.com/niecore/MoreUnit-Eclipse/pull/1

**Comment by niecore** [comment]
@apupier Thanks for the review. I'll fix the open points asap.

File: org.moreunit.core/src/org/moreunit/core/preferences/GenericConfigurationPage.java [hunk 9]
```diff
@@ -1,242 +1,141 @@
...
-
-    private class PageValidator implements ModifyListener
-    {
-        public void modifyText(ModifyEvent event)
-        {
-            validate();
-        }
-    }
-}
+package org.moreunit.core.preferences;
```
[the remark this answers is not in the document]
Comment (niecore) [reply org.moreunit.core/src/org/moreunit/core/preferences/GenericConfigurationPage.java]: I checked the file endings and both, the upstream and my branch, are using "CR" "LF" as line ending. Probably the diff-tool has problems here.

**Comment by Andrei-Pozolotin** [comment]
+1

**Comment by RoiSoleil** [comment]
Too old PR. Not merging.

**Pull request closed without merging**